packages/adblocker/src/engine/engine.ts

```diff
@@ -229,6 +233,122 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
+      patterns: Record<string, IPattern>;
+    } = {
+      organizations: {},
+      categories: {},
+      patterns: {},
+    };
+    const resources: Map<string, Resource> = new Map();
+
+    type ConfigKey = keyof {
+      [Key in keyof Config as Config[Key] extends boolean ? Key : never]: Config[Key];
+    };
+
+    const compatibleConfigKeys: ConfigKey[] = [];
+    const configKeysMustMatch: ConfigKey[] = (Object.keys(config) as (keyof Config)[]).filter(
+      function (key): key is ConfigKey {
+        return (
+          typeof config[key] === 'boolean' && !compatibleConfigKeys.includes(key as ConfigKey)
+        );
+      },
+    );
+
+    for (const engine of engines) {
+      // Validate the config
+      for (const configKey of configKeysMustMatch) {
+        if (config[configKey] !== engine.config[configKey]) {
+          throw new Error(`config "${configKey}" of all merged engines must be the same`);
+        }
+      }
+
+      const filters = engine.getFilters();
+
+      for (const networkFilter of filters.networkFilters) {
+        networkFilters.set(networkFilter.getId(), networkFilter);
+      }
+
+      for (const cosmeticFilter of filters.cosmeticFilters) {
+        cosmeticFilters.set(cosmeticFilter.getId(), cosmeticFilter);
+      }
+
+      for (const preprocessor of engine.preprocessors.preprocessors) {
+        preprocessors.push(preprocessor);
+      }
+
+      for (const [key, value] of engine.lists) {
+        if (lists.has(key)) {
+          continue;
+        }
+
+        lists.set(key, value);
+      }
+
+      if (engine.metadata !== undefined) {
+        for (const organization of engine.metadata.organizations.getValues()) {
+          if (metadata.organizations[organization.key] === undefined) {
+            metadata.organizations[organization.key] = organization;
+          }
+        }
+        for (const category of engine.metadata.categories.getValues()) {
+          if (metadata.categories[category.key] === undefined) {
+            metadata.categories[category.key] = category;
+          }
+        }
+        for (const pattern of engine.metadata.patterns.getValues()) {
+          if (metadata.patterns[pattern.key] === undefined) {
+            metadata.patterns[pattern.key] = pattern;
+          }
+        }
+      }
+
+      for (const [name, resource] of engine.resources.resources) {
+        if (!resources.has(name)) {
+          resources.set(name, resource);
+        }
+      }
+    }
+
+    const resourcesText = [...resources.entries()]
+      .reduce((state, [name, resource]) => {
+        return [...state, `${name} ${resource.contentType}\n${resource.body}`];
+      }, [] as string[])
+      .join('\n\n');
+
+    const engine = new this({
+      networkFilters: Array.from(networkFilters.values()),
+      cosmeticFilters: Array.from(cosmeticFilters.values()),
+      preprocessors,
+
+      lists,
+      config,
+    }) as InstanceType<T>;
+    engine.metadata = new Metadata(metadata);
```

Comment:
If Metadata was `undefined` in all engines I think it should still be undefined.

Reply:
Fixed in [`5cbe880` (#4138)](https://github.com/ghostery/adblocker/pull/4138/commits/5cbe8806e89807a5ace2e98ff6c226eb930bea04)